examples/vision/focal_modulation_network.py

```diff
@@ -0,0 +1,1220 @@
+        self.pi = tf.constant(np.pi)
+
+    def __call__(self, step):
+        if self.total_steps < self.warmup_steps:
+            raise ValueError("Total_steps must be larger or equal to warmup_steps.")
+        cos_annealed_lr = tf.cos(
+            self.pi
+            * (tf.cast(step, tf.float32) - self.warmup_steps)
+            / float(self.total_steps - self.warmup_steps)
+        )
+        learning_rate = 0.5 * self.learning_rate_base * (1 + cos_annealed_lr)
+        if self.warmup_steps > 0:
+            if self.learning_rate_base < self.warmup_learning_rate:
+                raise ValueError(
+                    "Learning_rate_base must be larger or equal to "
+                    "warmup_learning_rate."
+                )
+            slope = (
+                self.learning_rate_base - self.warmup_learning_rate
+            ) / self.warmup_steps
+            warmup_rate = slope * tf.cast(step, tf.float32) + self.warmup_learning_rate
+            learning_rate = tf.where(
+                step < self.warmup_steps, warmup_rate, learning_rate
+            )
+        return tf.where(
+            step > self.total_steps, 0.0, learning_rate, name="learning_rate"
+        )
+
+
+total_steps = int((len(x_train) / BATCH_SIZE) * EPOCHS)
+warmup_epoch_percentage = 0.15
+warmup_steps = int(total_steps * warmup_epoch_percentage)
+scheduled_lrs = WarmUpCosine(
+    learning_rate_base=LEARNING_RATE,
+    total_steps=total_steps,
+    warmup_learning_rate=0.0,
+    warmup_steps=warmup_steps,
+)
+
+"""
+## Initialize, compile and train the model
+"""
+
+focal_mod_net = FocalModulationNetwork()
+optimizer = AdamW(learning_rate=scheduled_lrs, weight_decay=WEIGHT_DECAY)
+
+# Compile and train the model.
+focal_mod_net.compile(
+    optimizer=optimizer,
+    loss="sparse_categorical_crossentropy",
+    metrics=["accuracy"],
+)
+history = focal_mod_net.fit(
+    train_ds,
+    epochs=EPOCHS,
+    validation_data=val_ds,
+    callbacks=callbacks,
+)
+
+"""
+## Plot loss and accuracy
+"""
+
+plt.plot(history.history["loss"], label="loss")
+plt.plot(history.history["val_loss"], label="val_loss")
+plt.legend()
+plt.show()
+
+plt.plot(history.history["accuracy"], label="accuracy")
+plt.plot(history.history["val_accuracy"], label="val_accuracy")
+plt.legend()
+plt.show()
+
+"""
+# Test Visulaizations
+
+Let's test our model on some test images and see how the gates look like.
+"""
+
+test_images, test_labels = next(iter(test_ds))
+_ = focal_mod_net(test_images)
+
+# Take the mid layer for visualization
+gates = focal_mod_net.basic_layers[1].blocks[-1].modulation.gates
+gates = upsampler(gates)
+modulator = focal_mod_net.basic_layers[1].blocks[-1].modulation.modulator
+modulator = upsampler(modulator)
+
+# Plot the test images with the gates and modulator overlayed.
+for row in range(5):
+    display_grid(
+        test_images=test_images,
+        gates=gates,
+        modulator=modulator,
+    )
+
+"""
+# Conclusion
```

Comment:
This example over all appears to be too long. Can it be refocused and shortened?

Reply:
We agree.

Also we would like to know which one you consider long here, the code or the explanations. We found a lot of caveats in the official code repo which were not spelled out in the paper, making our explanations long and verbose.

Please let us know how you would like us to move ahead.

Reply:
Having a lot of text is not a problem. But we should make sure to minimize the amount of code. The normal guideline is a limit of 300 lines of code (which does not include docstrings and comments).

Reply:
We have refactored the code. The code still goes over the 300 line mark, but we think the present state is necessary for readability.

If it is not a blocker, it would be great if we could move ahead with this.